src/realm/util/interprocess_condvar.hpp

```diff
@@ -115,28 +127,93 @@ class InterprocessCondVar {
 private:
     // non-zero if a shared part has been registered (always 0 on process local instances)
     SharedPart* m_shared_part = nullptr;
+
 #ifdef REALM_CONDVAR_EMULATION
     // keep the path to allocated system resource so we can remove them again
     std::string m_resource_path;
     // pipe used for emulation. When using a named pipe, m_fd_read is read-write and m_fd_write is unused.
     // When using an anonymous pipe (currently only for tvOS) m_fd_read is read-only and m_fd_write is write-only.
     int m_fd_read = -1;
     int m_fd_write = -1;
+#endif
 
 #ifdef _WIN32
-    // Semaphore used to queue up threads waiting for the condition to
-    // become signaled. 
-    HANDLE m_sema = 0;
-    // An auto-reset event used by the broadcast/signal thread to wait
-    // for all the waiting thread(s) to wake up and be released from the
-    // semaphore. 
-    HANDLE m_waiters_done = 0;
-    std::string m_name;
-
-    // Serialize access to m_waiters_count
-    InterprocessMutex m_waiters_lockcount;
-#endif
+    // A wrapper around HANDLE that auto-closes.
+    struct HandleHolder {
+        HandleHolder() = default;
+        /*implicit*/ HandleHolder(HANDLE h)
+            : handle(h)
+        {
+        }
+        ~HandleHolder()
+        {
+            if (handle)
+                REALM_ASSERT_RELEASE(CloseHandle(handle));
+        }
+        HandleHolder(HandleHolder&& other) noexcept
+            : handle(std::exchange(other.handle, {}))
+        {
+        }
+        HandleHolder& operator=(HandleHolder&& other) noexcept
+        {
+            if (handle)
+                REALM_ASSERT_RELEASE(CloseHandle(handle));
+            handle = std::exchange(other.handle, {});
+            return *this;
+        }
+
+        /*implicit*/ operator HANDLE() const
+        {
+            return handle;
+        }
+
+        explicit operator bool() const
+        {
+            return bool(handle);
+        }
+
+        HANDLE handle = {};
+    };
+
+    struct Event {
+        void wait(DWORD millis = INFINITE) noexcept;
+        void set() noexcept;
+        void reset() noexcept;
+        HandleHolder handle;
+    };
+
+    struct Mutex {
+        void lock() noexcept;
+        bool try_lock() noexcept;
+        void unlock() noexcept;
+        HandleHolder handle;
+    };
```

Comment:
Do we already have helpers for these? Or is there a better place to put these? Or is it best to keep at least `Event` and `Mutex` private here since most code should be using our cross-platform wrappers.?

Reply:
Deferring to separate PR